presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/StaticMetastoreLocator.java

```diff
@@ -101,15 +102,14 @@ public ThriftMetastoreClient createMetastoreClient(Optional<String> delegationTo
     {
         Comparator<Backoff> comparator = Comparator.comparingLong(Backoff::getBackoffDuration)
                 .thenComparingLong(Backoff::getLastFailureTimestamp);
-        List<Integer> indices = backoffs.stream()
+        List<Backoff> backoffsSorted = backoffs.stream()
                 .sorted(comparator)
-                .map(backoffs::indexOf)
                 .collect(toImmutableList());
 
         TException lastException = null;
-        for (int index : indices) {
+        for (Backoff backoff : backoffsSorted) {
             try {
-                return getClient(addresses.get(index), backoffs.get(index), delegationToken);
+                return getClient(backoff.getAddress(), backoff, delegationToken);
```

Comment:
Construct a list of attempted addresses (in order of attempt) so that "Failed connecting to Hive metastore: " below is accurate.

Then you can remove `addresses` field.

Reply:
Yeah - did not come to my mind that maybe order in error message is not that important.

Reply:
Actually unsorted list could still be recovered from List of backoffs.